File: Nikhil/python/react_agent.py
# react_agent.py
from __future__ import annotations

import logging
from typing import List, Sequence, TYPE_CHECKING

from pydantic import BaseModel, Field

from agent import Agent
from json_schema import JsonSchema
from steps import Step
from tool import Tool

if TYPE_CHECKING:  # avoid circular imports at runtime
    from executor_module import ExecutorModule
    from critic_module import CriticModule

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(name)s [%(levelname)s] %(message)s",
)
logger = logging.getLogger(__name__)


class ReactAgent(Agent):
    DEFAULT_MODEL: str = "gpt-4.1"

    class Parameters(BaseModel):
        thought: str = Field(
            ...,
            description="Your reasoning about why this tool has been called.",
        )

    # ------------------------------------------------------------------ #
    # construction
    # ------------------------------------------------------------------ #
    def __init__(
        self,
        id_: str,
        description: str,
        tools: Sequence[Tool],
        check_last_step: bool = True,
    ) -> None:
        super().__init__(id_=id_, description=description, tools=[])

        self.model = self.DEFAULT_MODEL
        self.temperature = 0.0

        self._context: str = ""
        self._examples: str = ""

        # list of execution steps moved here
        self._steps: List[Step] = []

        # inner modules
        from executor_module import ExecutorModule  # local import
        from critic_module import CriticModule

        self._executor: ExecutorModule = ExecutorModule(
            agent=self,
            tools=list(tools),
            check_last_step=check_last_step,
            model=self.DEFAULT_MODEL,
        )
        self._reviewer: CriticModule = CriticModule(
            agent=self,
            tools=list(tools),
            model=self.DEFAULT_MODEL,
        )

    # ------------------------------------------------------------------ #
    # context & examples
    # ------------------------------------------------------------------ #
    @property
    def context(self) -> str:
        return self._context

    @context.setter
    def context(self, value: str) -> None:
        if value is None:
            raise ValueError("context must not be None")
        self._context = value

    @property
    def examples(self) -> str:
        return self._examples

    @examples.setter
    def examples(self, value: str) -> None:
        if value is None:
            raise ValueError("examples must not be None")
        self._examples = value

    # ------------------------------------------------------------------ #
    # steps management (now owned by ReactAgent)
    # ------------------------------------------------------------------ #
    @property
    def steps(self) -> List[Step]:
        return self._steps

    def get_last_step(self) -> Step | None:
        return self._steps[-1] if self._steps else None

    def add_step(self, step: Step) -> None:
        self._steps.append(step)
        try:
            logger.info(JsonSchema.serialize(step))
        except Exception:
            logger.exception("Unable to serialise step for logging")

    # ------------------------------------------------------------------ #
    # inner modules (read-only)
    # ------------------------------------------------------------------ #
    @property
    def executor(self):  # -> ExecutorModule
        return self._executor

    @property
    def reviewer(self):  # -> CriticModule
        return self._reviewer

    # ------------------------------------------------------------------ #
    # delegate
    # ------------------------------------------------------------------ #
    def execute(self, command: str) -> Step:
        if command is None:
            raise ValueError("command must not be None")
        logger.info("Executing command: %s", command)
        return self._executor.execute(command)

File: Nikhil/java/src/main/java/com/infosys/small/react/ReactAgent.java
package com.infosys.small.react;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.annotation.JsonView;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.infosys.small.core.Agent;
import com.infosys.small.core.Tool;
import com.kjetland.jackson.jsonSchema.annotations.JsonSchemaDescription;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import lombok.Setter;

/**
 * This implements a generic ReAct agent.
 */
public class ReactAgent extends Agent {

	public static final String DEFAULT_MODEL = "gpt-4.1";
//	public static final String DEFAULT_MODEL = "gpt-4o";
//	public static final String DEFAULT_MODEL = "o3";

	/**
	 * Base class for {@link Tool} parameters for those tools that are available to
	 * a ReAct agent. They must all accept at least a thought parameter.
	 */
	@NoArgsConstructor
	public static class Parameters {

		@JsonProperty(required = true)
		@JsonPropertyDescription("Your reasoning about why this tool has been called.")
		public @NonNull String thought;
	}

	@NoArgsConstructor
	@JsonSchemaDescription("This represent the final execution step performed by a ReAct agent.")
	public static class Step {

		// ** Strictly ignore this when translating into Python
		public static class Views {
			public interface Compact {
			}

			public interface Complete extends Compact {
			}
		}

		public enum Status {
			IN_PROGRESS, COMPLETED, ERROR
		};

		@JsonPropertyDescription("If you finish the execution or you experience an unrecoverable error, set this to either COMPLETED or ERROR respectively.")
		@JsonView(Views.Compact.class)
		public Status status;

		// Do not remove it's OK it stays here
		@JsonPropertyDescription("The tool or agent that executed this step. This is provided automatically, so you do not need to output it.")
		@JsonView(Views.Compact.class)
		public @NonNull String actor;

		@JsonProperty(required = true)
		@JsonPropertyDescription("Your reasoning about why and how accomplish this step.")
		@JsonView(Views.Compact.class)
		public @NonNull String thought;

		@JsonProperty(required = true)
		@JsonPropertyDescription("Any additional data, like step outcomes, error messages, etc..")
		@JsonView(Views.Compact.class)
		public @NonNull String observation;

		// Private constructor to force use of the builder
		private Step(Builder builder) {
			this.status = builder.status;
			this.actor = Objects.requireNonNull(builder.actor, "id must not be null");
			this.thought = Objects.requireNonNull(builder.thought, "thought must not be null");
			this.observation = Objects.requireNonNull(builder.observation, "observation must not be null");
		}

		/**
		 * Builder for Step.
		 */
		public static class Builder {
			private Status status;
			private String actor;
			private String thought;
			private String observation;

			public Builder status(Status status) {
				this.status = status;
				return this;
			}

			public Builder actor(@NonNull String actor) {
				this.actor = actor;
				return this;
			}

			public Builder thought(@NonNull String thought) {
				this.thought = thought;
				return this;
			}

			public Builder observation(@NonNull String observation) {
				this.observation = observation;
				return this;
			}

			public Step build() {
				return new Step(this);
			}
		}
	}

	@NoArgsConstructor
	@JsonSchemaDescription("This extends execution step to cover function calls")
	public static class ToolCallStep extends Step {

		@JsonProperty(required = true)
		@JsonPropertyDescription("The action that was taken at this step. It is typically a tool invocation.")
		@JsonView(Views.Compact.class)
		public @NonNull String action;

		@JsonProperty(required = true, value = "action_input")
		@JsonPropertyDescription("Input for the action.")
		@JsonView(Views.Compact.class)
		public @NonNull String actionInput;

		@JsonProperty(required = true, value = "action_steps")
		@JsonPropertyDescription("In case the action for this step was delegated to another agent, this is the list of steps that agent performed to complete the action.")
		@JsonView(Views.Complete.class)
		public @NonNull List<Step> actionSteps;

		private ToolCallStep(Builder builder) {
			super(builder);
			this.action = Objects.requireNonNull(builder.action, "action must not be null");
			this.actionInput = Objects.requireNonNull(builder.actionInput, "actionInput must not be null");
			this.actionSteps = new ArrayList<>(builder.actionSteps);
		}

		/**
		 * Builder for ToolCallStep.
		 */
		public static class Builder extends Step.Builder {
			private String action;
			private String actionInput;
			public List<Step> actionSteps = new ArrayList<>();

			@Override
			public Builder status(Status status) {
				return (Builder) super.status(status);
			}

			@Override
			public Builder actor(@NonNull String actor) {
				return (Builder) super.actor(actor);
			}

			@Override
			public Builder thought(@NonNull String thought) {
				return (Builder) super.thought(thought);
			}

			@Override
			public Builder observation(@NonNull String observation) {
				return (Builder) super.observation(observation);
			}

			public Builder action(@NonNull String action) {
				this.action = action;
				return this;
			}

			public Builder actionInput(@NonNull String actionInput) {
				this.actionInput = actionInput;
				return this;
			}

			public Builder actionSteps(@NonNull List<? extends Step> steps) {
				this.actionSteps = new ArrayList<>(steps);
				return this;
			}

			public Builder addAllSteps(@NonNull List<? extends Step> steps) {
				this.actionSteps.addAll(steps);
				return this;
			}

			public Builder addStep(@NonNull Step step) {
				this.actionSteps.add(step);
				return this;
			}

			@Override
			public ToolCallStep build() {
				return new ToolCallStep(this);
			}
		}
	}

	/**
	 * Any additional context you want to provide to the agent.
	 */
	@Getter
	@Setter
	private @NonNull String context = "";

	/**
	 * Any additional examples you want to provide to the agent.
	 */
	@Getter
	@Setter
	private @NonNull String examples = "";

	/**
	 * The list of steps executed so far while executing current command.
	 */
	@Getter
	private final @NonNull List<Step> steps = new ArrayList<>();

	/**
	 * @return Last step so far (or null).
	 */
	public Step getLastStep() {
		return (steps.size() == 0) ? null : steps.get(steps.size() - 1);
	}

	/**
	 * Adds one step to the list of execution steps.
	 */
	void addStep(Step step) {
		steps.add(step);

		// TODO Write step into the database, if a link to the database is provided
	}

	@Getter(AccessLevel.PROTECTED)
	private final @NonNull ExecutorModule executor;

	@Getter(AccessLevel.PROTECTED)
	private final @NonNull CriticModule reviewer;

	// TODO Make it a wrapper of another agent instead? A lot of code forwarding...

	public ReactAgent(@NonNull String id, @NonNull String description, @NonNull List<? extends Tool> tools,
			boolean checkLastStep) {

		// Only executor is using tools
		super(id, description, new ArrayList<>());
		setModel(DEFAULT_MODEL);
		setTemperature(0d);

		this.executor = new ExecutorModule(this, tools, checkLastStep, DEFAULT_MODEL);
		this.reviewer = new CriticModule(this, tools, DEFAULT_MODEL);
	}

	public Step execute(@NonNull String command) throws JsonProcessingException {
		return executor.execute(command);
	}
}
